File: FE/src/app/services/user.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { GetUserDataResponse } from '../dtos/response/GetUserDataResponse';
import { UpdateUserDataRequest } from '../dtos/request/UpdateUserDataRequest';
import { OrganizzatoreResponse } from '../dtos/response/OrganizzatoreResponse';
import { GetOrganizzatoriSeguiti } from '../dtos/response/GetOrganizzatoriSeguiti';
import { MessageResponse } from '../dtos/response/MessageResponse';

/**
 * service per gestire le interazioni col backend riguardo l'utente.
 */
@Injectable({
  providedIn: 'root'
})
export class UserService {
  private backendUrl: string = 'http://localhost:8080/api/v1/user/';

  //costruttore dove istanzio le classi con cui interagire
  constructor(private http: HttpClient) { }

  /**
   * metodo per prendere i dati di un utente
   * @returns DTO con i dati dell'utente richiesto
   */
  getUserData(): Observable<GetUserDataResponse> {
    const header = this.getHeader();

    return this.http.get<GetUserDataResponse>(this.backendUrl + 'getUserData/' + localStorage.getItem('id')?.toString(), { headers: header });
  }

  /**
   * metodo per aggiornare i propri dati
   * @param nome nome dell'utente
   * @param cognome cognome dell'utente
   * @param email email dell'utente
   * @param username username dell'utente
   * @param vecchiaPassword vecchia password dell'utente
   * @param nuovaPassword nuova password dell'utente
   * @param iscrittoNewsletter configurazione della newsletter dell'utente
   * @param oldUsername vecchio username dell'utente
   * @returns DTO con i dati aggiornati dell'utente
   */
  updateUserData(nome: string, cognome: string, email: string, username: string, vecchiaPassword: string, nuovaPassword: string, iscrittoNewsletter: boolean): Observable<GetUserDataResponse> {
    const header = this.getHeader();
    const request: UpdateUserDataRequest = { nome, cognome, email, username, vecchiaPassword, nuovaPassword, iscrittoNewsletter };

    return this.http.put<GetUserDataResponse>(this.backendUrl + 'updateUserData/' + localStorage.getItem('id')?.toString(), request, { headers: header });
  }

  /**
   * metodo per l'admin per aggiornare i dati di un utente
   * @param nome nome dell'utente
   * @param cognome cognome dell'utente
   * @param email email dell'utente
   * @param username username dell'utente
   * @param vecchiaPassword vecchia password dell'utente
   * @param nuovaPassword nuova password dell'utente
   * @param iscrittoNewsletter configurazione della newsletter dell'utente
   * @param oldUsername vecchio username dell'utente
   * @returns DTO con i dati aggiornati dell'utente
   */
  adminUpdateUserData(nome: string, cognome: string, email: string, username: string, vecchiaPassword: string, nuovaPassword: string, iscrittoNewsletter: boolean, oldUsername: string): Observable<GetUserDataResponse> {
    const header = this.getHeader();
    //creo il DTO con i dati da aggiorare
    const request: UpdateUserDataRequest = { nome, cognome, email, username, vecchiaPassword, nuovaPassword, iscrittoNewsletter };

    return this.http.put<GetUserDataResponse>(this.backendUrl + 'adminUpdateUserData/' + oldUsername, request, { headers: header });
  }

  /**
   * metodo per eliminare l'account di un utente
   * @returns messaggio di risposta
   */
  eliminaAccount(): Observable<MessageResponse> {
    const header = this.getHeader();

    return this.http.delete<MessageResponse>(this.backendUrl + 'delete/' + localStorage.getItem('id')?.toString(), { headers: header });
  }

  /**
   * chiamo il backend per prendere i dati di un utente dato un username
   * @param username username dell'utente
   * @returns DTO con i dati dell'utente
   */
  getAdminUserData(username: string): Observable<GetUserDataResponse> {
    const header = this.getHeader();

    return this.http.get<GetUserDataResponse>(this.backendUrl + 'getAdminUserData/' + username, { headers: header });
  }

  /**
   * chiamo il backend per eliminare i dati di un altro utente (solo se si è admin)
   * @param username username dell'utente in questione
   * @returns messaggio di risposta
   */
  adminEliminaUser(username: string): Observable<MessageResponse> {
    const header = this.getHeader();

    return this.http.delete<MessageResponse>(this.backendUrl + 'adminDelete/' + username, { headers: header });
  }

  /**
   * metodo per prendere tutti gli organizzatori salvati sul database
   * @returns lista di DTO con i dati degli organizzatori
   */
  getAllOrganizzatori(): Observable<OrganizzatoreResponse[]> {
    const header = this.getHeader();

    return this.http.get<OrganizzatoreResponse[]>(this.backendUrl + 'getAllOrganizzatori', { headers: header });
  }

  /**
   * metodo per far seguire ad un turista un organizzatore
   * @param organizzatoreId id dell'organizzatore da seguire
   * @returns messaggio di risposta
   */
  seguiOrganizzatore(organizzatoreId: string): Observable<MessageResponse> {
    const header = this.getHeader();

    return this.http.get<MessageResponse>(this.backendUrl + 'seguiOrganizzatore/' + organizzatoreId + '/' + localStorage.getItem('id')?.toString().trim(), { headers: header });
  }

  /**
   * metodo per far smettere di seguire ad un turista un organizzatore
   * @param organizzatoreId id dell'organizzatore da unfolloware
   * @returns messaggio di risposta
   */
  smettiSeguireOrganizzatore(organizzatoreId: string): Observable<MessageResponse> {
    const header = this.getHeader();

    return this.http.get<MessageResponse>(this.backendUrl + 'smettiSeguireOrganizzatore/' + organizzatoreId + '/' + localStorage.getItem('id')?.toString().trim(), { headers: header });
  }

  /**
   * metodo per prendere tutti gli organizzatori seguiti da un dato turista
   * @returns array di DTO con gli organizzatori seguiti
   */
  getAllOrganizzatoriSeguiti(): Observable<GetOrganizzatoriSeguiti[]> {
    const header = this.getHeader();

    return this.http.get<GetOrganizzatoriSeguiti[]>(this.backendUrl + 'getOrganizzatoriSeguiti/' + localStorage.getItem('id')?.toString().trim(), { headers: header });
  }

  //creo l'header con il token da mandare al backend
  private getHeader(): HttpHeaders {
    return new HttpHeaders({
      'Authorization': localStorage.getItem('token') ? `${localStorage.getItem('token')}` : '',
      id: localStorage.getItem('id') ? `${localStorage.getItem('id')}` : '',
      ruolo: localStorage.getItem('ruolo') ? `${localStorage.getItem('ruolo')}` : ''
    });
  }
}

File: BE/src/main/java/it/polimi/iswpf/service/_interface/UserService.java
package it.polimi.iswpf.service._interface;

import it.polimi.iswpf.dto.request.UpdateUserDataRequest;
import it.polimi.iswpf.dto.response.OrganizzatoreResponse;
import it.polimi.iswpf.dto.response.OrganizzatoreSeguitoResponse;
import it.polimi.iswpf.dto.response.UserResponse;
import it.polimi.iswpf.service.implementation.UserServiceImpl;

import java.util.List;

/**
 * Interfaccia che contiene le firme dei metodi del service inerenti all'user.
 * Implementazione -> {@link UserServiceImpl}.
 */
public interface UserService {

    /**
     * Metodo che, dato un id, prende l'utente dal database tramite la repository e lo ritorna al controller.
     * @param userId Id dell'utente.
     * @return DTO con tutti i dati dell'utente.
     */
    UserResponse getUserData(Long userId);

    /**
     * Metodo che, modifica i dati dell'utente con quell'id.
     * @param userId Id dell'utente.
     * @param request DTO con i nuovi dati.
     * @return DTO con i dati dell'utente aggiornati.
     */
    UserResponse updateUserData(Long userId, UpdateUserDataRequest request);

    /**
     * Metodo che permette all'admin di modificare i dati di un utente, dato un username.
     * @param username Username dell'utente da modificare.
     * @param request DTO con i nuovi dati.
     * @return DTO con i dati dell'utente aggiornati.
     */
    UserResponse adminUpdateUserData(String username, UpdateUserDataRequest request);

    /**
     * Elimina l'utente dal database dopo aver fatto diversi controlli.
     * @param userId Id dell'utente da eliminare.
     */
    void deleteAccount(Long userId);

    /**
     * Dato un username, ed eseguiti diversi controlli, ritorno l'utente con quell'username.
     * @param username Username dell'utente richiesto.
     * @return DTO con i dati dell'utente richiesto.
     */
    UserResponse getAdminUserData(String username);

    /**
     * Dato un username, viene eliminato l'account.
     * @param username Username dell'account da eliminare.
     */
    void adminDeleteAccount(String username);

    /**
     * Metodo per prendere tutti gli organizzatori presenti sul database.
     * @return Lista di DTO con i dati di ogni organizzatore.
     */
    List<OrganizzatoreResponse> getAllOrganizzatori();

    /**
     * Metodo per prendere tutti gli utenti di un dato ruolo.
     * @param ruolo Ruolo richiesto, passato in modo dinamico tramite l'endpoint.
     * @return Lista di DTO con i dati degli utenti richiesti.
     */
    List<UserResponse> getAllByRuolo(String ruolo);

    /**
     * Metodo che permette a un turista di seguire un organizzatore, e quindi essere notificati alla creazione di un evento.
     * @param organizzatoreId Id univoco dell'organizzatore, passato in modo dinamico tramite l'endpoint.
     * @param turistaId Id univoco del turista, passato in modo dinamico tramite l'endpoint.
     */
    void seguiOrganizzatore(Long organizzatoreId, Long turistaId);

    /**
     * Metodo che, dato un turista, restituisce gli username degli organizzatori seguiti.
     * @param turistaId Id univoco del turista, passato in modo dinamico tramite l'endpoint.
     * @return Lista di DTO con gli username degli organizzatori.
     */
    List<OrganizzatoreSeguitoResponse> getOrganizzatoriSeguiti(Long turistaId);

    /**
     * Metodo che permette a un turista di smettere di seguire un organizzatore.
     * @param organizzatoreId Id univoco dell'organizzatore, passato in modo dinamico tramite l'endpoint.
     * @param turistaId Id univoco del turista, passato in modo dinamico tramite l'endpoint.
     */
    void smettiSeguireOrganizzatore(Long organizzatoreId, Long turistaId);
}
